File: core/src/main/scala/org/apache/spark/sql/delta/commands/DeletionVectorUtils.scala
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.delta.commands

import org.apache.spark.sql.delta.{DeletionVectorsTableFeature, DeltaConfigs, Snapshot, SnapshotDescriptor}
import org.apache.spark.sql.delta.actions.{Metadata, Protocol}
import org.apache.spark.sql.delta.files.SupportsRowIndexFilters
import org.apache.spark.sql.delta.files.TahoeFileIndex
import org.apache.spark.sql.delta.sources.DeltaSQLConf

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.execution.datasources.FileIndex
import org.apache.spark.sql.functions.col
import org.apache.spark.util.Utils

trait DeletionVectorUtils {

  /**
   * Run a query on the delta log to determine if the given snapshot contains no deletion vectors.
   * Return `false` if it does contain deletion vectors.
   */
  def isTableDVFree(spark: SparkSession, snapshot: Snapshot): Boolean = {
    val dvsReadable = deletionVectorsReadable(snapshot)

    if (dvsReadable) {
      val dvCount = snapshot.allFiles
        .filter(col("deletionVector").isNotNull)
        .limit(1)
        .count()

      dvCount == 0L
    } else {
      true
    }
  }

  /**
   * Returns true if persistent deletion vectors are enabled and
   * readable with the current reader version.
   */
  def fileIndexSupportsReadingDVs(fileIndex: FileIndex): Boolean = fileIndex match {
    case index: TahoeFileIndex => deletionVectorsReadable(index)
    case _: SupportsRowIndexFilters => true
    case _ => false
  }

  def deletionVectorsWritable(
      snapshot: SnapshotDescriptor,
      newProtocol: Option[Protocol] = None,
      newMetadata: Option[Metadata] = None): Boolean =
    deletionVectorsWritable(
      protocol = newProtocol.getOrElse(snapshot.protocol),
      metadata = newMetadata.getOrElse(snapshot.metadata))

  def deletionVectorsWritable(protocol: Protocol, metadata: Metadata): Boolean =
    protocol.isFeatureSupported(DeletionVectorsTableFeature) &&
      DeltaConfigs.ENABLE_DELETION_VECTORS_CREATION.fromMetaData(metadata)

  def deletionVectorsReadable(
      snapshot: SnapshotDescriptor,
      newProtocol: Option[Protocol] = None,
      newMetadata: Option[Metadata] = None): Boolean = {
    deletionVectorsReadable(
      newProtocol.getOrElse(snapshot.protocol),
      newMetadata.getOrElse(snapshot.metadata))
  }

  def deletionVectorsReadable(
      protocol: Protocol,
      metadata: Metadata): Boolean = {
    protocol.isFeatureSupported(DeletionVectorsTableFeature) &&
      metadata.format.provider == "parquet" // DVs are only supported on parquet tables.
  }

  /**
   * Utility method that checks the table has no Deletion Vectors enabled. Deletion vectors
   * are supported in read-only mode for now. Any updates to tables with deletion vectors
   * feature are disabled until we add support.
   */
  def assertDeletionVectorsNotReadable(
      spark: SparkSession, metadata: Metadata, protocol: Protocol): Unit = {
    val disable =
      Utils.isTesting && // We are in testing and enabled blocking updates on DV tables
          spark.conf.get(DeltaSQLConf.DELTA_ENABLE_BLOCKING_UPDATES_ON_DV_TABLES)
    if (!disable && deletionVectorsReadable(protocol, metadata)) {
      throw new UnsupportedOperationException(
        "Updates to tables with Deletion Vectors feature enabled are not supported in " +
          "this version of Delta Lake.")
    }
  }

  /**
   * Utility method that checks the table metadata has no deletion vectors enabled. Deletion vectors
   * are supported in read-only mode for now. Any updates to metadata to enable deletion vectors are
   * blocked until we add support.
   */
  def assertDeletionVectorsNotEnabled(
    spark: SparkSession, metadata: Metadata, protocol: Protocol): Unit = {
    val disable =
      Utils.isTesting && // We are in testing and enabled blocking updates on DV tables
        spark.conf.get(DeltaSQLConf.DELTA_ENABLE_BLOCKING_UPDATES_ON_DV_TABLES)
    if (!disable &&
      (protocol.isFeatureSupported(DeletionVectorsTableFeature) ||
        DeltaConfigs.ENABLE_DELETION_VECTORS_CREATION.fromMetaData(metadata)
      )
    ) {
      throw new UnsupportedOperationException(
        "Enabling Deletion Vectors on the table is not supported in this version of Delta Lake.")
    }
  }
}

// To access utilities from places where mixing in a trait is inconvenient.
object DeletionVectorUtils extends DeletionVectorUtils

File: spark/v2/src/main/java/io/delta/spark/internal/v2/read/deletionvector/DeletionVectorUtils.java
/*
 * Copyright (2025) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.delta.spark.internal.v2.read.deletionvector;

import io.delta.kernel.internal.actions.DeletionVectorDescriptor;
import io.delta.kernel.internal.actions.Metadata;
import io.delta.kernel.internal.actions.Protocol;
import io.delta.kernel.internal.tablefeatures.TableFeatures;
import java.util.Optional;
import org.apache.spark.sql.delta.DeltaParquetFileFormat;
import org.apache.spark.sql.delta.RowIndexFilterType;
import org.apache.spark.sql.types.StructType;
import scala.Option;
import scala.Tuple2;
import scala.collection.immutable.Map;
import scala.collection.immutable.Map$;

/**
 * Utility class for deletion vector support in the V2 connector.
 *
 * <p>Provides methods for:
 *
 * <ul>
 *   <li>Checking if a table supports deletion vectors
 *   <li>Converting Kernel DV descriptors to Spark format
 *   <li>Building DV metadata for PartitionedFile
 *   <li>Schema augmentation with DV columns
 * </ul>
 */
public final class DeletionVectorUtils {

  private DeletionVectorUtils() {}

  /** Check if table supports reading deletion vectors. */
  public static boolean isReadable(Protocol protocol, Metadata metadata) {
    return protocol.supportsFeature(TableFeatures.DELETION_VECTORS_RW_FEATURE)
        && "parquet".equalsIgnoreCase(metadata.getFormat().getProvider());
  }

  /** Convert Kernel's DeletionVectorDescriptor to Spark's DeletionVectorDescriptor. */
  public static org.apache.spark.sql.delta.actions.DeletionVectorDescriptor convertToSpark(
      DeletionVectorDescriptor kernelDv) {
    return new org.apache.spark.sql.delta.actions.DeletionVectorDescriptor(
        kernelDv.getStorageType(),
        kernelDv.getPathOrInlineDv(),
        Option.apply(kernelDv.getOffset().orElse(null)),
        kernelDv.getSizeInBytes(),
        kernelDv.getCardinality(),
        Option.empty());
  }

  /**
   * Build metadata map containing DV descriptor if present.
   *
   * <p>The metadata is stored in PartitionedFile.otherConstantMetadataColumnValues and used by
   * DeltaParquetFileFormat to generate the is_row_deleted column.
   */
  @SuppressWarnings("unchecked")
  public static Map<String, Object> buildMetadata(Optional<DeletionVectorDescriptor> dvOpt) {
    if (!dvOpt.isPresent()) {
      return (Map<String, Object>) (Map<?, ?>) Map$.MODULE$.empty();
    }

    String dvBase64 = convertToSpark(dvOpt.get()).serializeToBase64();

    return (Map<String, Object>)
        (Map<?, ?>)
            Map$.MODULE$
                .empty()
                .$plus(
                    new Tuple2<>(
                        DeltaParquetFileFormat.FILE_ROW_INDEX_FILTER_ID_ENCODED(), dvBase64))
                .$plus(
                    new Tuple2<>(
                        DeltaParquetFileFormat.FILE_ROW_INDEX_FILTER_TYPE(),
                        RowIndexFilterType.IF_CONTAINED));
  }

  /** Add the __delta_internal_is_row_deleted column to the schema. */
  public static StructType augmentSchema(StructType schema) {
    return schema.add(DeltaParquetFileFormat.IS_ROW_DELETED_STRUCT_FIELD());
  }
}
